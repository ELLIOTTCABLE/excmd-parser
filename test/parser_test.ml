open Excmd_parser

let scrpt desc str =
   Printf.printf "## %s:\nParser.script %S\n" desc str;
   AST.pp (Parser.script_of_string str);
   Printf.printf "\n\n"

let stmt desc str =
   Printf.printf "## %s:\nParser.statement %S\n" desc str;
   Statement.pp (Parser.statement_of_string str);
   Printf.printf "\n\n"


let tests () =
   (* Basic statements *)
   stmt "Simple, single command"
   "test";

   stmt "Single command, with count"
   "2test";

   (* Basic statements, with parameters *)
   stmt "Single command, with single positional parameter"
   "test foo";

   stmt "Single command, count, and single positional parameter"
   "2test foo";

   stmt "Single command, with multiple positional parameters"
   "test foo bar";

   stmt "Single command, count, and multiple positional parameters"
   "2test foo bar";

   stmt "Single command, with single boolean flag"
   "test --foo";

   stmt "Single command, with multiple boolean flags"
   "test --foo --bar";

   stmt "Single command, with single boolean short-flag"
   "test -f";

   stmt "Single command, with multiple, concatenated boolean short-flags"
   "test -foo";

   stmt "Single command, with multiple, separated boolean short-flags"
   "test -f -o -o";

   stmt "Single command, with single possibly-parameterized flag"
   "test --foo bar";

   stmt "Single command, with single possibly-parameterized short-flag"
   "test -f bar";

   stmt "Single command with single possibly-parameterized flag followed by a positional parameter"
   "test --foo bar baz";

   stmt "Single command with two possibly-parameterized flags"
   "test --foo bar --baz widget";

   stmt "Single command, with single explicitly-parameterized flag"
   "test --foo=bar";

   stmt "Single command with single explicitly-parameterized flag followed by a positional parameter"
   "test --foo=bar baz";

   stmt "Single command with mixed flags and parameters"
   "test --foo bar --baz=widget qux -qu ux";

   (* Simple multi-statement scripts *)
   scrpt "Statements separated by semicolons"
   "test; 2test; 3test";

   scrpt "Statements separated by semicolons, with a trailing semicolon"
   "test; 2test; 3test;";

   scrpt "Statements, with arguments, separated by semicolons"
   "test --foo bar; 2test --foo=bar; 3test --foo bar";

   scrpt "Statements, with arguments, separated by semicolons, with a trailing semicolon"
   "test --foo bar; 2test --foo=bar; 3test --foo bar;";

   scrpt "Newlines after statements"
   "test;
   2test;
   3test";

   scrpt "Newlines after statements, with a trailing newline"
   "test;
   2test;
   3test;
   "


let () =
   match Sys.argv with

   (* Automated tests *)
   | [| _ |] -> tests ()

   (* Interactive usage *)
   | [| _; "script"; str |] -> AST.pp (Parser.script_of_string str)
   | [| _; "statement"; str |] -> Statement.pp (Parser.statement_of_string str)
   | [| _; _; _ |] -> raise (Invalid_argument
      "First argument must be a valid non-terminal (e.g. 'script' or 'statement')")

   | _ ->
      Printf.eprintf
      "!! Please provide either no arguments (to run the automated tests), or exactly two arguments (for\n";
      Printf.eprintf
      "!! interactive experimentation): a nonterminal entry-point (e.g. 'script'), and a string to parse.\n";
      raise (Invalid_argument "Either exactly two, or zero, arguments are required")
