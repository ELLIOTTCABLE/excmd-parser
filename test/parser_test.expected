

### Simple, single command:
Parser.statement "test"
{ "count": 1, "cmd": "test", "args": [] }

### Single command, with count:
Parser.statement "2test"
{ "count": 2, "cmd": "test", "args": [] }

### Single command, with single positional parameter:
Parser.statement "test foo"
{ "count": 1, "cmd": "test", "args": [ [ "Positional", "foo" ] ] }

### Single command, with single positional URL:
Parser.statement "test google.com/search"
{
  "count": 1,
  "cmd": "test",
  "args": [ [ "Positional", "google.com/search" ] ]
}

### Single command, count, and single positional parameter:
Parser.statement "2test foo"
{ "count": 2, "cmd": "test", "args": [ [ "Positional", "foo" ] ] }

### Single command, with multiple positional parameters:
Parser.statement "test foo bar"
{
  "count": 1,
  "cmd": "test",
  "args": [ [ "Positional", "foo" ], [ "Positional", "bar" ] ]
}

### Single command, count, and multiple positional parameters:
Parser.statement "2test foo bar"
{
  "count": 2,
  "cmd": "test",
  "args": [ [ "Positional", "foo" ], [ "Positional", "bar" ] ]
}

### Single command, with single boolean flag:
Parser.statement "test --foo"
{
  "count": 1,
  "cmd": "test",
  "args": [ [ "Flag", { "name": "foo", "payload": [ "Absent" ] } ] ]
}

### Single command, with multiple boolean flags:
Parser.statement "test --foo --bar"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "bar", "payload": [ "Absent" ] } ]
  ]
}

### Single command, with single boolean short-flag:
Parser.statement "test -f"
{
  "count": 1,
  "cmd": "test",
  "args": [ [ "Flag", { "name": "f", "payload": [ "Absent" ] } ] ]
}

### Single command, with multiple, concatenated boolean short-flags:
Parser.statement "test -foo"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "f", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "o", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "o", "payload": [ "Absent" ] } ]
  ]
}

### Single command, with multiple, separated boolean short-flags:
Parser.statement "test -f -o -o"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "f", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "o", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "o", "payload": [ "Absent" ] } ]
  ]
}

### Single command, with single possibly-parameterized flag:
Parser.statement "test --foo bar"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
    [ "Positional", "bar" ]
  ]
}

### Single command, with single possibly-parameterized short-flag:
Parser.statement "test -f bar"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "f", "payload": [ "Unresolved" ] } ],
    [ "Positional", "bar" ]
  ]
}

### Single command, single possibly-parameterized short-flag, and a URL:
Parser.statement "test -f google.com/search"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "f", "payload": [ "Unresolved" ] } ],
    [ "Positional", "google.com/search" ]
  ]
}

### Single command with single possibly-parameterized flag followed by a positional parameter:
Parser.statement "test --foo bar baz"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
    [ "Positional", "bar" ],
    [ "Positional", "baz" ]
  ]
}

### Single command with two possibly-parameterized flags:
Parser.statement "test --foo bar --baz widget"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
    [ "Positional", "bar" ],
    [ "Flag", { "name": "baz", "payload": [ "Unresolved" ] } ],
    [ "Positional", "widget" ]
  ]
}

### Single command, with single explicitly-parameterized flag:
Parser.statement "test --foo=bar"
{
  "count": 1,
  "cmd": "test",
  "args": [ [ "Flag", { "name": "foo", "payload": [ "Resolved", "bar" ] } ] ]
}

### Single command with single explicitly-parameterized flag followed by a positional parameter:
Parser.statement "test --foo=bar baz"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Resolved", "bar" ] } ],
    [ "Positional", "baz" ]
  ]
}

### Single command, single explicitly-parameterized flag with a URL payload, followed by a positional parameter:
Parser.statement "test --foo=google.com/search baz"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [
      "Flag",
      { "name": "foo", "payload": [ "Resolved", "google.com/search" ] }
    ],
    [ "Positional", "baz" ]
  ]
}

### Single command with mixed flags and parameters:
Parser.statement "test --foo bar --baz=widget qux -qu ux"
{
  "count": 1,
  "cmd": "test",
  "args": [
    [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
    [ "Positional", "bar" ],
    [ "Flag", { "name": "baz", "payload": [ "Resolved", "widget" ] } ],
    [ "Positional", "qux" ],
    [ "Flag", { "name": "q", "payload": [ "Absent" ] } ],
    [ "Flag", { "name": "u", "payload": [ "Unresolved" ] } ],
    [ "Positional", "ux" ]
  ]
}

### Statements separated by semicolons:
Parser.script "test; 2test; 3test"
{
  "statements": [
    { "count": 1, "cmd": "test", "args": [] },
    { "count": 2, "cmd": "test", "args": [] },
    { "count": 3, "cmd": "test", "args": [] }
  ]
}

### Statements separated by semicolons, with a trailing semicolon:
Parser.script "test; 2test; 3test;"
{
  "statements": [
    { "count": 1, "cmd": "test", "args": [] },
    { "count": 2, "cmd": "test", "args": [] },
    { "count": 3, "cmd": "test", "args": [] }
  ]
}

### Statements, with arguments, separated by semicolons:
Parser.script "test --foo bar; 2test --foo=bar; 3test --foo bar"
{
  "statements": [
    {
      "count": 1,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
        [ "Positional", "bar" ]
      ]
    },
    {
      "count": 2,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Resolved", "bar" ] } ]
      ]
    },
    {
      "count": 3,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
        [ "Positional", "bar" ]
      ]
    }
  ]
}

### Statements, with arguments, separated by semicolons, with a trailing semicolon:
Parser.script "test --foo bar; 2test --foo=bar; 3test --foo bar;"
{
  "statements": [
    {
      "count": 1,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
        [ "Positional", "bar" ]
      ]
    },
    {
      "count": 2,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Resolved", "bar" ] } ]
      ]
    },
    {
      "count": 3,
      "cmd": "test",
      "args": [
        [ "Flag", { "name": "foo", "payload": [ "Unresolved" ] } ],
        [ "Positional", "bar" ]
      ]
    }
  ]
}

### Newlines after statements:
Parser.script "test;\n   2test;\n   3test"
{
  "statements": [
    { "count": 1, "cmd": "test", "args": [] },
    { "count": 2, "cmd": "test", "args": [] },
    { "count": 3, "cmd": "test", "args": [] }
  ]
}

### Newlines after statements, with a trailing newline:
Parser.script "test;\n   2test;\n   3test;\n   "
{
  "statements": [
    { "count": 1, "cmd": "test", "args": [] },
    { "count": 2, "cmd": "test", "args": [] },
    { "count": 3, "cmd": "test", "args": [] }
  ]
}

### An acceptable statement, incrementally:
Incremental.statement "hello --where=world"
{
  "count": 1,
  "cmd": "hello",
  "args": [
    [ "Flag", { "name": "where", "payload": [ "Resolved", "world" ] } ]
  ]
}

### A failing statement, incrementally:
Incremental.statement "hello --where="
fail-continuation invoked! cool.


### Listing of acceptable tokens during a failure:
Incremental.statement "hello --where="
URL_START, QUOTE_OPEN, IDENTIFIER


### Dumping of stack-debugging information from a checkpoint:
Incremental.statement "hello --where="
InputNeeded
Terminal
unit
EQUALS
EQUALS : (unit) Terminal
FLAG_LONG : (string) Terminal
command : (string) Nonterminal
option_COUNT_ : (string option) Nonterminal
list_COLON_ : (unit list) Nonterminal


### Ascertaining of current command from a checkpoint:
Incremental.statement "a_command_name --blah="
a_command_name
